plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2055,15 +2069,23 @@ void producer_plugin_impl::produce_block() {
       return sigs;
    } );
 
-   chain.commit_block();
+   if ( blockvault_plug && blockvault_plug->get() != nullptr ) {
+      std::promise<bool> p;
+      std::future<bool> f = p.get_future();
+      std::optional<producer_watermark> watermark{get_watermark(pending_blk_state->header.producer)};
+      EOS_ASSERT(watermark.has_value(), empty_watermark, "Attempting to use a watermark that does not exist");
+      blockvault_plug->get()->async_propose_constructed_block(watermark.value(), pending_blk_state->dpos_irreversible_blocknum, pending_blk_state->block, [&p](bool b) {
+         p.set_value( b );
+      });
+      EOS_ASSERT( f.get(), blockvault_failure, "Blockvault failure" );
```

Comment:
This isn't so much blockvault failure as it is invalid block. Seems like blockvault could report an error code indicating why the block was not accepted and that could be logged and included in the exception.
A blockvault failure would be if you f.wait_for() some duration. The timeout would be an indication that the blockvault failed. If blockvault was out of process you really should have a timeout. As is, probably not needed.

Reply:
Added a TODO for making a more granular approach to failures on the blockvault server-side.